graphql_compiler/compiler/workarounds/orientdb_query_execution.py

```diff
@@ -0,0 +1,341 @@
+        preferred_locations.add(first_match_step.as_block.location)
+    else:
+        eligible_locations.add(first_match_step.as_block.location)
+
+    # This loop will repeat the analysis of the first step of the first traversal.
+    # QueryRoots other than the first are required to always be at a location whose status
+    # (preferred / eligible / ineligible) is already known. Since we already processed
+    # the first QueryRoot above, the rest of the loop can assume all QueryRoots are like that.
+    for current_traversal in match_query.match_traversals:
+        for match_step in current_traversal:
+            current_step_location = match_step.as_block.location
+
+            already_encountered_location = any((
+                current_step_location in preferred_locations,
+                current_step_location in eligible_locations,
+                current_step_location in ineligible_locations,
+            ))
+
+            if isinstance(match_step.root_block, QueryRoot):
+                if not already_encountered_location:
+                    raise AssertionError(u'Unexpectedly encountered a location in QueryRoot whose '
+                                         u'status has not been determined: {} {} {}'
+                                         .format(current_step_location, match_step, match_query))
+
+                at_eligible_location = (
+                    current_step_location in preferred_locations or
+                    current_step_location in eligible_locations)
+
+                # This location has already been encountered and processed.
+                # Other than setting the "at_eligible_location" state for the sake of
+                # the following MATCH steps, there is nothing further to be done.
+                continue
+            elif isinstance(match_step.root_block, Recurse):
+                # All Recurse blocks cause locations within to be ineligible.
+                at_eligible_location = False
+            elif isinstance(match_step.root_block, Traverse):
+                # Optional Traverse blocks cause locations within to be ineligible.
+                # Non-optional Traverse blocks do not change the eligibility of locations within:
+                # if the pre-Traverse location was eligible, so will the location within,
+                # and if it was not eligible, neither will the location within.
+                if match_step.root_block.optional:
+                    at_eligible_location = False
+            else:
+                raise AssertionError(u'Unreachable condition reached: {} {} {}'
+                                     .format(match_step.root_block, match_step, match_query))
+
+            if not at_eligible_location:
+                ineligible_locations.add(current_step_location)
+            elif match_step.where_block is not None and _is_local_filter(match_step.where_block):
```

Comment:
Why not move the `where_block is not None` logic inside the helper function, and rename it something like `_contains_local_filter`? This is repeated above (before the for loop).

Reply:
This is getting refactored -- the `where_block is not None` is now a separate conditional from the local filter check.